### ISSUE TITLE
pmap on pytree (namedtuple) with various dimensions

### ISSUE THREAD

**Issue by john-heyer** [google/jax#3102]
Hello! I have a named tuple that works fine with jitted functions. I'd then like to use vmap (or pmap) to a list of these named tuples. I see that the default behavior is that jax will push the `in_axes` argument down onto each of the named_tuple members. This brings up errors like "vmap got arg 2 of rank 0 but axis to be mapped 0" because of course we don't intend to map over the members of the named tup, but over the list of named tups itself.  For example:

```
testtup = namedtuple("testtup", ['x', 'y', 'n'])
tup_list = [testtup(x=np.ones(5)*i, y=np.zeros(5) + i, n=i) for i in range(4)]

def add_tup(t):
    return np.sum(t.x + t.y) * t.n

s = jit(add_tup)(l[0])  # works!
s_all = vmap(add_tup)(l)  # doesn't work :/
```

I have seen [this issue](https://github.com/google/jax/issues/2540), but didn't think it was the same. My workaround is to just make the members of the namedtup 2d arrays (or 1d arrays for scalars). Is this by design? Thanks!

**Comment by skye** [comment]
`vmap` and `pmap` only map over arrays, not arbitrary Python sequences, so your workaround sounds like the right approach. This is by design: the underlying tracing infrastructure that powers jax's function transformations are based on array operations, and we automatically traverse other Python types to gather the constituent arrays for the transforms. You can learn more about this in the docs: [JAX pytrees](https://jax.readthedocs.io/en/latest/notebooks/JAX_pytrees.html)

**Comment by mattjj** [comment]
Well put, @skye! Indeed `vmap` vectorizes over array axes; without an axis of a contiguous array, i.e. if we just have a list of arrays instead, we can't vectorize things.

If you just want to map over a list, you can use Python's builtin `map`. If you want to concatenate and vectorize, as you mentioned in your workaround, one way to write it is with a helper function like this:

```python
from collections import namedtuple
from jax import vmap, vjp
from jax import tree_util
import jax.numpy as np

testtup = namedtuple("testtup", ['x', 'y', 'n'])
tup_list = [testtup(x=np.ones(5)*i, y=np.zeros(5) + i, n=i) for i in range(4)]

def add_tup(t):
  return np.sum(t.x + t.y) * t.n


def stackmap(f, lst):
  stacked = tree_util.tree_multimap(lambda *args: np.stack(args), *lst)
  out_stacked = vmap(f)(stacked)
  _, outer_treedef = tree_util.tree_flatten([object()] * len(lst))
  _, inner_treedef = tree_util.tree_flatten(out_stacked)
  out_unstacked_transposed = tree_util.tree_map(list, out_stacked)
  out_unstacked = tree_util.tree_transpose(outer_treedef, inner_treedef,
                                           out_unstacked_transposed)
  return out_unstacked

s_all = stackmap(add_tup, tup_list)
```

Since I think @skye pretty much answered the question (whether JAX comes built in with a way of vectorized-mapping over lists) and this helper function shows a way to use `vmap` after some concatenation/splitting, I'm going to close this issue. Please open a new one as more questions arise!

**Comment by john-heyer** [comment]
Thank you! All very helpful 😄
